app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/endpoints.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/keycode_state_changed.h>
+#include <zmk/events/modifiers_state_changed.h>
+#include <zmk/hid.h>
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+enum mode {
+    MODE_KEY_UP = 0,
+    MODE_KEY_DOWN = 1,
+    MODE_HOLD = 2,
+};
+
+struct behavior_macro_config {
+    int mode;
+    int sleep;
+    int behavior_count;
+    struct zmk_behavior_binding *behaviors;
+};
+
+static int behavior_macro_init(const struct device *dev) { return 0; }
+
+static void press_and_release_combo(const struct behavior_macro_config *cfg,
+                                    struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_pressed(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+        behavior_keymap_binding_released(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
+    }
+}
+
+static void press_combo(const struct behavior_macro_config *cfg,
+                        struct zmk_behavior_binding_event event) {
+    for (int index = 0; index < cfg->behavior_count; index++) {
+        const struct device *behavior = device_get_binding(cfg->behaviors[index].behavior_dev);
+        if (!behavior) {
+            break;
+        }
+        behavior_keymap_binding_pressed(&cfg->behaviors[index], event);
+        k_msleep(cfg->sleep);
```

Comment:
In the event that you set `sleep = 0` to press multiple things at the same time, I don't think we should call k_msleep. A sleep of 0 may be treated as a yield (see zephyr/kernel/sched.c `z_tick_sleep()`)